File: cpp/sampler.cpp
#include <sampler.h>

using namespace std;

Sampler::Sampler(const map<Cell, float> &visited,
        const Size &size,
        const float &average,
        const int &stdev,
        const int &maxDepth) :
    _visited(visited), _size(size),
    _average(average), _stdev(stdev), _maxDepth(maxDepth) {
    _boundary = getBoundary(visited);
}

float Sampler::get_or_inf(Cell cell) {
    if(_visited.count(cell)) {
        return _visited[cell];
    } else if(_sampled.count(cell)) {
        return _sampled[cell];
    } else
        return -INF;
}

float Sampler::get(Cell cell) {
    if(_visited.count(cell)) {
        return _visited[cell];
    } else if(_sampled.count(cell)) {
        return _sampled[cell];
    } else
        return _average;
}

vector<Cell> Sampler::getBoundary(const map<Cell, float> &cells) {
    vector<Cell> boundary;

    for(auto cellvalpair : cells) {
        auto cell= cellvalpair.first;

        bool isBoundary = false;

        vector<Cell> nbroffs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        for(auto nbroff : nbroffs) {
            auto nbr = addDelta(cell, nbroff);
            if(cells.count(nbr) == 0) {
                isBoundary = true;
                break;
            }
        }

        if(isBoundary) {
            boundary.push_back(cell);
        }
    }

    return boundary;
}

void Sampler::sample() {

    _sampled.clear();

    PriorityQueue<Cell> pq;

    for(auto cell : _boundary) {
        pq.update(cell, 0);
    }

    while(true) {
        auto cellpriopair = pq.findMin();
        auto cell = cellpriopair.first;
        auto prio = cellpriopair.second;

        if(prio < 0 || prio >= _maxDepth) {
            break;
        }

        vector<Cell> nbroffs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        for(auto nbroff : nbroffs) {
            auto nbr = addDelta(cell, nbroff);
            if(_visited.count(nbr) == 0 && _sampled.count(nbr) == 0) {
                if(isValid(nbr, _size)) {
                    pq.update(nbr, prio+1);
                }
            }
        }

        if(_visited.count(cell))
            continue;

        vector<float> nbrvals;

        for(auto nbroff : nbroffs) {
            auto nbr = addDelta(cell, nbroff);
            if(get_or_inf(nbr) >= 0) {
                nbrvals.push_back(get_or_inf(nbr));
            }
        }

        assert(nbrvals.size() > 0);

        random_shuffle(nbrvals.begin(), nbrvals.end());
        auto randomnbrval = nbrvals[0];

        float randomnoise = (drand48() - 0.5)*_stdev;
        auto cellval = randomnbrval + randomnoise;
        if(cellval < 0)
            cellval = 0;

        _sampled[cell] = cellval;
    }
}

File: sampler.py
import util
import random

class Sampler:
    def __init__(self, visited, size, average=30, stdev=20, maxDepth=50):
        self.visited = visited
        self.size = size
        self.average = average
        self.stdev = stdev
        self.boundary = self.getBoundary(self.visited.keys())
        self.maxDepth = maxDepth

    # return the value of the cell if it is sampled or visited, else return None
    def get_or_none(self, cell):
        if cell in self.visited:
            return self.visited[cell]
        elif cell in self.sampled:
            return self.sampled[cell]
        else:
            return None

    def get(self, cell):
        if self.get_or_none(cell) != None:
            return self.get_or_none(cell)
        return self.average

    def getBoundary(self, cells):
        boundary = []

        for cell in cells:
            isBoundary = False
            for nbroff in [(0,1), (0,-1), (1,0), (-1,0)]:
                nbr = tuple([x+i for x,i in zip(cell, nbroff)])
                if nbr not in cells:
                    isBoundary = True
            if isBoundary:
                boundary.append(cell)

        return boundary

    def sample(self):
        # first clear all the sampled cells, we are generating a new sample
        self.sampled = {}

        pq = util.PriorityQueue()

        # push the boundary cells onto the priority queue, this is the base case
        for b in self.boundary:
            pq.update(b, 0)

        while True:
            cell, prio = pq.removeMin()

            # termination conditions
            if cell == None or prio >= self.maxDepth:
                break

            # add all cells to the queue which have not been sampled or visited
            for nbroff in [(0,1), (0,-1), (1,0), (-1,0), (1,1), (-1,-1), (1,-1),
                          (-1,1)]:
                nbr = util.addDelta(cell, nbroff)
                if nbr not in self.sampled and nbr not in self.visited:
                    if util.isValid(nbr, self.size):
                        pq.update(nbr, prio+1)

            # if cell is already visited, no need to sample this cell
            if cell in self.visited:
                continue

            # get all the neighbor values which have been either visited or
            # sampled
            nbrvals = []
            for nbroff in [(0,1), (0,-1), (1,0), (-1,0), (1,1), (-1,-1), (1,-1),
                          (-1,1)]:
                nbr = util.addDelta(cell, nbroff)
                if self.get_or_none(nbr) != None:
                    nbrvals.append(self.get_or_none(nbr))

            assert len(nbrvals) > 0

            # draw a random neighbor and draw my value as a gaussian around that
            nbrval = random.sample(nbrvals, 1)[0]
            cellval = random.gauss(nbrval, self.stdev)
            if cellval < 0:
                cellval = 0

            self.sampled[cell] = cellval


if __name__ == '__main__':
    visited = {
        (0, 0): 0.6,
        (0, 1): 0.6,
        (0, 2): 0.6,
        (0, 3): 0.6,
        (0, 4): 0.6,
        (1, 0): 0.6,
        (1, 1): 0.6,
        (1, 2): 0.6,
        (1, 3): 0.6,
        (1, 4): 0.6,
        (2, 0): 0.6,
        (2, 1): 0.6,
        (2, 2): 0.6,
        (2, 3): 0.6,
        (2, 4): 0.6,
        (3, 0): 0.6,
        (3, 1): 0.6,
        (3, 2): 0.6,
        (3, 3): 0.6,
        (3, 4): 0.6,
        (4, 0): 0.6,
        (4, 1): 0.6,
        (4, 2): 0.6,
        (4, 3): 0.6,
        (4, 4): 0.6,
    }
    sp = Sampler(visited, (100, 100))
    print sp.boundary

    for _ in range(1000):
        sp.sample()
    print sp.sampled
